datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs

```diff
@@ -224,6 +226,62 @@ impl<S: SimplifyInfo> ExprSimplifier<S> {
     }
 }
 
+#[allow(rustdoc::private_intra_doc_links)]
+/// Canonicalize any BinaryExprs that are not in canonical form
+/// <literal> <op> <col> is rewritten to <col> <op> <literal> (remember to switch the operator)
+/// <col> <op> <literal> is canonical
+/// <col1> <op> <col2> is rewritten so that the name of col1 sorts higher than col2 (b > a would be canonicalized to a < b)
+struct Canonicalizer {}
+
+impl Canonicalizer {
+    fn new() -> Self {
+        Self {}
+    }
+}
+
+impl TreeNodeRewriter for Canonicalizer {
+    type N = Expr;
+
+    fn mutate(&mut self, expr: Expr) -> Result<Expr> {
+        if let Expr::BinaryExpr(BinaryExpr { left, op, right }) = expr {
+            // Case 1, <col1> <op> <col2>
+            let mut new_expr = BinaryExpr {
+                left: left.clone(),
+                op: op.clone(),
+                right: right.clone(),
+            };
```

Comment:
I wonder if there is a cleaner way to do this, without having to clone here (since this will always clone for `Expr;:BinaryExpr` even if we don't end up swapping)

Could try an approach without relying on mutability of `new_expr`

Reply:
I use `return` in `if` statement to avoid `new_expr` in the new changes :)